### ISSUE TITLE
High memory usage from monitoring_thread_loop

### ISSUE THREAD

**Issue by timwoj** [mheily/libkqueue#153]
We discovered a problem recently where the monitoring thread spawned by libkqueue was using a very large amount of memory. After a bit of digging, we realized that it's allocating two arrays of `int`s that are the size of the maximum number of file descriptors on the system. The system we discovered this on is a Rocky 9 system, with a docker container running in containerd. In this environment, containerd sets the maximum number of descriptors to 1073741816. libkqueue then allocates about 8GB of memory.

While it is possible to tune the container down to use fewer descriptors, this seems like an upstream problem in libkqueue itself.

**Comment by william-stearns** [comment]
This issue caused Zeek, the above-mentioned use of libjqueue, to grow from <300M of memory used to 4.2GB of resident ram (16GB virtual) when run on a system that has this large number of file descriptors (which appears to be RHEL 9/centos stream 9/alma 9/ and rocky 9.  Is it possible to adjust the size of these arrays to avoid having zeek allocate this much ram, triggering the oom-killer?

**Comment by timwoj** [comment]
We fixed this partially by explicitly capping the number of open files in our runtime environment, but also in a PR in a libkqueue fork that avoids calling `calloc` on every fd entry. See https://github.com/zeek/libkqueue/pull/2. I can open a PR over here if that's something you guys think is useful.

**Comment by arr2036** [comment]
Yeah seems like a reasonable fix.  Just add comments about _why_ it's done this way, so people don't accidentally undo it in future.
